lib/internal/url.js

```diff
@@ -1284,7 +1284,8 @@ function urlToOptions(url) {
     options.port = Number(url.port);
   }
   if (url.username || url.password) {
-    options.auth = `${url.username}:${url.password}`;
+    options.auth =
+      `${decodeURIComponent(url.username)}:${decodeURIComponent(url.password)}`;
```

Comment:
This introduces a spec violation. From [RFC 7617, section 2](https://tools.ietf.org/html/rfc7617#section-2):

     Furthermore, a user-id containing a colon character is invalid, as
     the first colon in a user-pass string separates user-id and password
     from one another; text after the first colon is part of the password.
     User-ids containing colons cannot be encoded in user-pass strings.

With this change, I can now write something like:
```js
https.get('https://not%3Agood:god@example.com')
```
And the server will interpret that as user="not" and pass="good:god" instead of pass="god" (which is the password we still all use for our root accounts, right? Right!)

Reply:
But what would be proper behavior in case of passing `:` in username? Encoding username - the wrong way for the computation Authorization header. Pre-validation with throwing an exception? Maybe...

Reply:
I don’t really want to introduce new exceptions here … maybe just leave it encoded?

The same question also pops up for e.g. `%2F` for `/` in the password.

Reply:
Yes. I think `:` and `/` are the only problematic ones; for `@` it's "last at-sign wins."

It's probably okay to leave those encoded, that's no worse from how it was before.

Reply:
@bnoordhuis Done, PTAL